wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift

```diff
@@ -0,0 +1,240 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+extension Swift.Result {
+    func flatMapAsync<NewSuccess>(_ transform: (Success) async -> Swift.Result<NewSuccess, Failure>) async -> Swift.Result<NewSuccess, Failure> {
+        switch self {
+        case .success(let success):
+            return await transform(success)
+        case .failure(let failure):
+            return .failure(failure)
+        }
+    }
+}
+
+public protocol HttpClient {
+
+    func send(_ request: ZMTransportRequest) async -> ZMTransportResponse
+
+}
+
+public class HttpClientImpl: HttpClient {
+
+    let transportSession: TransportSessionType
+    let queue: ZMSGroupQueue
+
+    public init(transportSession: TransportSessionType, queue: ZMSGroupQueue) {
+        self.transportSession = transportSession
+        self.queue = queue
+    }
+
+    public func send(_ request: ZMTransportRequest) async -> ZMTransportResponse {
+        await withCheckedContinuation { continuation in
+            request.add(ZMCompletionHandler(on: queue, block: { response in
+                continuation.resume(returning: response)
+            }))
+
+            transportSession.enqueueOneTime(request)
+        }
+    }
+}
+
+enum MessageTarget {
+    case Users(usersIds: [QualifiedID])
+    case Clients(recipients: [QualifiedClientID])
+    case Conversation(usersToIgnore: Set<QualifiedID>)
+}
+
+enum MessageSendError: Error {
+    case messageProtocolMissing
+    case unresolvedApiVersion
+    case failedToGenerateRequest
+    case failedToParseResponse
+    case gaveUpRetrying
+}
+
+typealias Message = ProteusMessage & MLSMessage
+
+public class MessageSender {
+
+    public init (
+        httpClient: HttpClient,
+        clientRegistrationDelegate: ClientRegistrationDelegate,
+        sessionEstablisher: SessionEstablisher,
+        context: NSManagedObjectContext
+    ) {
+        self.httpClient = httpClient
+        self.clientRegistrationDelegate = clientRegistrationDelegate
+        self.sessionEstablisher = sessionEstablisher
+        self.managedObjectContext = context
+    }
+
+    private let httpClient: HttpClient
+    private let managedObjectContext: NSManagedObjectContext
+    private let clientRegistrationDelegate: ClientRegistrationDelegate
+    private let sessionEstablisher: SessionEstablisher
+    private let requestFactory = ClientMessageRequestFactory()
+    private let processor = MessageSendingStatusPayloadProcessor()
+
+    func sendMessage(message: any Message) async -> Swift.Result<Void, MessageSendError> {
```

Comment:
I don't see why we use Result. We could throw the Error instead, no?

Reply:
Yes, we can. Different schools and I was used chaining result functions together.

I'm open to to changing to to throwing errors instead.

Reply:
The nice thing about `Result` is that the error is concrete so you know what the possible errors are. @netbe do you have a preference for using throwing functions?

Reply:
I refactored using try/catch in https://github.com/wireapp/wire-ios/pull/648

I'd tend to agree that it made the logic more straight forward since the need to chain calls mostly goes away.